web3/_utils/abi.py

```diff
@@ -145,26 +152,168 @@ def validate_value(cls, value):
         super().validate_value(value)
 
 
-class AcceptsHexStrMixin:
+class AcceptsHexStrEncoder(encoding.BaseEncoder):
+    def __init__(self, subencoder):
+        self.subencoder = subencoder
+
+    @property
+    def is_dynamic(self):
+        return self.subencoder.is_dynamic
+
+    @classmethod
+    def from_type_str(cls, abi_type, registry):
+        subencoder_cls = cls.get_subencoder_class()
+        subencoder = subencoder_cls.from_type_str(abi_type, registry)
+        return cls(subencoder)
+
+    @classmethod
+    def get_subencoder_class(cls):
+        if cls.subencoder_cls is None:
+            raise AttributeError(f'No subencoder class is set. {cls.__name__}')
+        return cls.subencoder_cls
+
+    @combomethod
     def validate_value(self, value):
+        normalized_value = self.validate_and_normalize(value)
+        return self.subencoder.validate_value(normalized_value)
+
+    def encode(self, value):
+        normalized_value = self.validate_and_normalize(value)
+        return self.subencoder.encode(normalized_value)
+
+    def validate_and_normalize(self, value):
+        raw_value = value
         if is_text(value):
             try:
                 value = decode_hex(value)
             except binascii.Error:
                 self.invalidate_value(
                     value,
-                    msg='invalid hex string',
+                    msg=f'{value} is an invalid hex string',
+                )
+            else:
+                if raw_value[:2] != '0x':
+                    if self.is_strict:
+                        self.invalidate_value(
+                            raw_value,
+                            msg='hex string must be prefixed with 0x'
+                        )
+                    elif raw_value[:2] != '0x':
+                        warnings.warn(
+                            'in v6 it will be invalid to pass a hex string without the "0x" prefix',
+                            category=DeprecationWarning
+                        )
+        return value
+
+
+class BytesEncoder(AcceptsHexStrEncoder):
+    subencoder_cls = encoding.BytesEncoder
+    is_strict = False
+
+
+class ByteStringEncoder(AcceptsHexStrEncoder):
+    subencoder_cls = encoding.ByteStringEncoder
+    is_strict = False
+
+
+class StrictByteStringEncoder(AcceptsHexStrEncoder):
+    subencoder_cls = encoding.ByteStringEncoder
+    is_strict = True
+
+
+class ExactLengthBytesEncoder(encoding.BaseEncoder):
+    # TODO: move this to eth-abi once the api is stabilized
```

Comment:
Might be good to open a tracking issue for this so it doesn't get lost.  Probably a "Good First Issue" candidate.